File: sdk/python/shark_auth/vault.py
"""Shark Token Vault client — fetch fresh 3rd-party OAuth credentials,
disconnect vault connections (with optional agent cascade), and issue
DPoP-authenticated token fetches.

W2 Method 9 adds:
- VaultDisconnectResult
- VaultTokenResult
- VaultClient.disconnect()
- VaultClient.fetch_token()
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import List, Optional

from . import _http
from .dpop import DPoPProver
from .errors import VaultError


@dataclass
class VaultToken:
    """A fresh, server-refreshed 3rd-party access token from the vault."""

    access_token: str
    expires_at: Optional[int]
    provider: Optional[str]
    scopes: List[str]


# ---------------------------------------------------------------------------
# W2 Method 9 — new typed results
# ---------------------------------------------------------------------------


@dataclass
class VaultDisconnectResult:
    """Result of disconnecting a vault connection.

    Attributes
    ----------
    connection_id:
        The connection that was deleted.
    revoked_agent_ids:
        Agent IDs whose tokens were cascade-revoked (empty when
        ``cascade_to_agents=False``).
    revoked_token_count:
        Total OAuth tokens revoked across all cascade-revoked agents.
    cascade_audit_event_id:
        Audit event ID for the cascade operation, or ``None`` when no
        cascade occurred.
    """

    connection_id: str
    revoked_agent_ids: List[str]
    revoked_token_count: int
    cascade_audit_event_id: Optional[str] = None


@dataclass
class VaultTokenResult:
    """A vault-brokered token returned via a DPoP-authenticated agent request.

    Attributes
    ----------
    access_token:
        The decrypted 3rd-party access token.
    token_type:
        Token type, typically ``"Bearer"``.
    expires_at:
        ISO-8601 expiry string, or ``None`` if the server did not supply one.
    provider:
        Provider name (e.g. ``"google_gmail"``).
    """

    access_token: str
    token_type: str = "Bearer"
    expires_at: Optional[str] = None
    provider: Optional[str] = None


class VaultClient:
    """Shark Token Vault client — connection management + delegated token fetch.

    Parameters
    ----------
    base_url:
        Base URL of the SharkAuth server (e.g. ``"https://auth.example.com"``).
    admin_key:
        Admin API key (``sk_live_...``). Required for admin operations such
        as :meth:`disconnect`.
    http_client:
        Optional pre-configured :class:`requests.Session`.

    Backward Compatibility
    ----------------------
    The legacy constructor signature ``VaultClient(auth_url=..., access_token=...)``
    is preserved: ``auth_url`` maps to ``base_url`` and ``access_token`` maps to the
    bearer token used by :meth:`get_fresh_token`.

    Example
    -------
    >>> vault = VaultClient(base_url="https://auth.example.com", admin_key="sk_live_...")
    >>> result = vault.disconnect("conn_abc123", cascade_to_agents=True)
    >>> print(result.revoked_agent_ids)
    """

    def __init__(
        self,
        base_url: str = "",
        admin_key: Optional[str] = None,
        http_client=None,
        *,
        # Legacy kwargs kept for backward compat
        auth_url: str = "",
        access_token: str = "",
        session: Optional[object] = None,
        connections_path: str = "/admin/vault/connections",
    ) -> None:
        # Resolve base_url from new or legacy kwarg
        self._base = (base_url or auth_url).rstrip("/")
        self._admin_key = admin_key or access_token
        self._session = http_client or session or _http.new_session()
        self._connections_path = connections_path

    # ------------------------------------------------------------------
    # Helpers
    # ------------------------------------------------------------------

    def _admin_headers(self) -> dict:
        if self._admin_key:
            return {"Authorization": f"Bearer {self._admin_key}"}
        return {}

    def _headers(self) -> dict:
        """Alias for backward compat (legacy get_fresh_token path)."""
        return self._admin_headers()

    # ------------------------------------------------------------------
    # Legacy method — kept for backward compat
    # ------------------------------------------------------------------

    def get_fresh_token(self, connection_id: str) -> VaultToken:
        """REMOVED — there is no admin-key endpoint that returns a fresh
        3rd-party access token by ``connection_id``.

        The historical default path ``/admin/vault/connections/{id}/token``
        was never mounted by the backend; only ``DELETE
        /api/v1/admin/vault/connections/{id}`` exists for admin scope (see
        ``internal/api/router.go``).  Token retrieval is intentionally gated
        behind a DPoP-bound agent token so the vault can validate the
        ``cnf.jkt`` binding before decrypting the stored credential.

        Migration
        ---------
        Use :meth:`fetch_token` instead — request a DPoP-bound bearer via
        :meth:`OAuthClient.get_token_with_dpop` (scope ``vault:read``), then
        call ``vault.fetch_token(provider=..., bearer_token=..., prover=...)``
        to retrieve the decrypted 3rd-party token.

        Raises
        ------
        VaultError:
            Always — this method is intentionally unreachable.
        """
        raise VaultError(
            "VaultClient.get_fresh_token has been removed: no admin-key endpoint "
            "exists for this operation. Use VaultClient.fetch_token() with a "
            "DPoP-bound agent token (scope 'vault:read') instead. See the "
            "method docstring for the migration recipe."
        )

    # ------------------------------------------------------------------
    # W2 Method 9 — disconnect (with cascade)
    # ------------------------------------------------------------------

    def disconnect(
        self,
        connection_id: str,
        *,
        cascade_to_agents: bool = True,
    ) -> VaultDisconnectResult:
        """Disconnect a vault connection.

        If ``cascade_to_agents=True`` (default), also revokes tokens for any
        agent that has ever fetched from this vault connection (Layer 5 cascade).

        Wraps ``DELETE /api/v1/vault/connections/{id}``. Sends
        ``?cascade=true`` query parameter when cascading.

        Parameters
        ----------
        connection_id:
            The ``conn_*`` identifier of the vault connection to delete.
        cascade_to_agents:
            When ``True`` (default), the server cascade-revokes tokens for
            all agents that accessed this connection. Pass ``False`` to
            disconnect silently without agent impact.

        Returns
        -------
        VaultDisconnectResult
            Identifiers of cascade-revoked agents, total token count, and
            the optional cascade audit event ID.

        Raises
        ------
        VaultError:
            On 4xx/5xx responses.

        Example
        -------
        >>> result = vault.disconnect("conn_abc123")
        >>> print(result.revoked_agent_ids, result.revoked_token_count)
        """
        params = {"cascade": "true" if cascade_to_agents else "false"}
        url = f"{self._base}/api/v1/vault/connections/{connection_id}"
        resp = _http.request(
            self._session, "DELETE", url, headers=self._admin_headers(), params=params
        )

        if resp.status_code in (200, 204):
            if resp.status_code == 204 or not resp.content:
                return VaultDisconnectResult(
                    connection_id=connection_id,
                    revoked_agent_ids=[],
                    revoked_token_count=0,
                )
            body = resp.json()
            return VaultDisconnectResult(
                connection_id=body.get("connection_id", connection_id),
                revoked_agent_ids=body.get("revoked_agent_ids", []),
                revoked_token_count=body.get("revoked_token_count", 0),
                cascade_audit_event_id=body.get("cascade_audit_event_id"),
            )

        if resp.status_code == 404:
            raise VaultError(
                f"vault connection not found: {connection_id}", status_code=404
            )
        if resp.status_code == 401:
            raise VaultError("not authorized (401)", status_code=401)
        if resp.status_code == 403:
            raise VaultError("forbidden — admin key required (403)", status_code=403)
        raise VaultError(
            f"disconnect failed: HTTP {resp.status_code}: {resp.text[:200]}",
            status_code=resp.status_code,
        )

    # ------------------------------------------------------------------
    # W2 Method 9 — fetch_token (DPoP-authenticated)
    # ------------------------------------------------------------------

    def fetch_token(
        self,
        *,
        provider: str,
        bearer_token: str,
        prover: DPoPProver,
    ) -> VaultTokenResult:
        """Fetch a fresh OAuth token from the vault using a DPoP-bound agent token.

        ``GET /api/v1/vault/{provider}/token`` with
        ``Authorization: DPoP <bearer_token>`` and a DPoP proof header.
        The server validates the DPoP proof, ``vault:read`` scope, and the
        ``tok.UserID`` binding before decrypting and returning the stored token.

        Parameters
        ----------
        provider:
            Provider slug, e.g. ``"google_gmail"``.
        bearer_token:
            The agent's DPoP-bound access token (issued by SharkAuth).
        prover:
            The :class:`~shark_auth.DPoPProver` instance whose private key
            matches the ``cnf.jkt`` bound to ``bearer_token``.

        Returns
        -------
        VaultTokenResult
            Decrypted 3rd-party access token, type, expiry, and provider name.

        Raises
        ------
        VaultError:
            On any 4xx/5xx response.

        Example
        -------
        >>> prover = DPoPProver.generate()
        >>> token = client.oauth.get_token_with_dpop(
        ...     grant_type="client_credentials",
        ...     dpop_prover=prover,
        ...     client_id="shark_agent_...",
        ...     client_secret="...",
        ...     scope="vault:read",
        ... )
        >>> result = vault.fetch_token(
        ...     provider="google_gmail",
        ...     bearer_token=token.access_token,
        ...     prover=prover,
        ... )
        >>> print(result.access_token)
        """
        url = f"{self._base}/api/v1/vault/{provider}/token"
        proof = prover.make_proof(htm="GET", htu=url)
        headers = {
            "Authorization": f"DPoP {bearer_token}",
            "DPoP": proof,
        }
        resp = _http.request(self._session, "GET", url, headers=headers)

        if resp.status_code == 200:
            body = resp.json()
            return VaultTokenResult(
                access_token=body["access_token"],
                token_type=body.get("token_type", "Bearer"),
                expires_at=body.get("expires_at"),
                provider=body.get("provider", provider),
            )

        if resp.status_code == 401:
            raise VaultError("invalid or expired DPoP token (401)", status_code=401)
        if resp.status_code == 403:
            raise VaultError("missing vault:read scope (403)", status_code=403)
        if resp.status_code == 404:
            raise VaultError(
                f"no vault connection for provider '{provider}' (404)", status_code=404
            )
        raise VaultError(
            f"vault fetch_token failed: HTTP {resp.status_code}: {resp.text[:200]}",
            status_code=resp.status_code,
        )

File: sdk/typescript/src/vault.ts
/**
 * Shark Token Vault client — fetch fresh 3rd-party OAuth credentials.
 *
 * @example
 * ```ts
 * const vault = new VaultClient({
 *   authUrl: "https://auth.example",
 *   accessToken: agentToken,
 * });
 * const fresh = await vault.exchange("conn_abc");
 * console.log(fresh.accessToken);
 * ```
 */

import { httpRequest } from "./http.js";
import { VaultError } from "./errors.js";

/** A fresh, server-refreshed 3rd-party access token from the vault. */
export interface VaultToken {
  accessToken: string;
  expiresAt?: number;
  provider?: string;
  scopes: string[];
}

/** Result of disconnecting a vault connection. */
export interface VaultDisconnectResult {
  connection_id: string;
  revoked_agent_ids: string[];
  revoked_token_count: number;
  cascade_audit_event_id?: string;
}

/** A vault connection record returned by listConnections. */
export interface VaultConnection {
  id: string;
  provider?: string;
  user_id?: string;
  created_at?: string;
  [key: string]: unknown;
}

/** Response from listConnections. */
export interface VaultConnectionListResult {
  data: VaultConnection[];
  total: number;
}

/** Options for listConnections. */
export interface ListConnectionsOptions {
  limit?: number;
  offset?: number;
  user_id?: string;
}

/** Options for constructing a {@link VaultClient}. */
export interface VaultClientOptions {
  /** Base URL of the SharkAuth server. */
  authUrl: string;
  /** Bearer access token for the calling agent. */
  accessToken: string;
  /**
   * Callback invoked when the server returns 401.
   * Should return a fresh access token. Called at most `maxRetries` times.
   */
  onRefresh?: () => Promise<string>;
  /** Maximum number of 401-retry attempts. Default: 2. */
  maxRetries?: number;
  /** Override the connections path. Default: `/admin/vault/connections`. */
  connectionsPath?: string;
  /** Admin API key for admin-scoped operations (disconnect, listConnections). */
  adminKey?: string;
}

/**
 * Fetches auto-refreshed 3rd-party tokens from the Shark Token Vault.
 *
 * On a `401` response, {@link VaultClient} will call the user-supplied
 * `onRefresh` callback (if any) to get a fresh agent token and retry the
 * request. Retries are bounded by `maxRetries` (default 2) to prevent
 * infinite loops.
 */
export class VaultClient {
  private readonly _authUrl: string;
  private _accessToken: string;
  private readonly _onRefresh?: () => Promise<string>;
  private readonly _maxRetries: number;
  private readonly _connectionsPath: string;
  private readonly _adminKey?: string;

  constructor(opts: VaultClientOptions) {
    this._authUrl = opts.authUrl.replace(/\/+$/, "");
    this._accessToken = opts.accessToken;
    this._onRefresh = opts.onRefresh;
    this._maxRetries = opts.maxRetries ?? 2;
    this._connectionsPath = opts.connectionsPath ?? "/admin/vault/connections";
    this._adminKey = opts.adminKey;
  }

  private _adminAuth(): Record<string, string> {
    const key = this._adminKey ?? this._accessToken;
    return { Authorization: `Bearer ${key}` };
  }

  /**
   * @deprecated Use {@link fetchToken} instead — same shape, preferred name matching Python SDK.
   */
  async exchange(referenceToken: string): Promise<VaultToken> {
    return this.fetchToken(referenceToken);
  }

  private async _exchangeWithRetry(
    connectionId: string,
    attempt: number
  ): Promise<VaultToken> {
    const url = `${this._authUrl}${this._connectionsPath}/${connectionId}/token`;
    const resp = await httpRequest(url, {
      method: "GET",
      headers: { Authorization: `Bearer ${this._accessToken}` },
    });

    if (resp.status === 200) {
      let body: Record<string, unknown>;
      try {
        body = resp.json<Record<string, unknown>>();
      } catch {
        throw new VaultError(
          `vault response is not valid JSON: ${resp.text.slice(0, 200)}`
        );
      }

      let scopes: string[];
      const rawScopes = body["scopes"] ?? body["scope"];
      if (typeof rawScopes === "string") {
        scopes = rawScopes.split(/\s+/).filter(Boolean);
      } else if (Array.isArray(rawScopes)) {
        scopes = rawScopes.map(String);
      } else {
        scopes = [];
      }

      return {
        accessToken: String(body["access_token"]),
        expiresAt:
          typeof body["expires_at"] === "number" ? body["expires_at"] : undefined,
        provider:
          typeof body["provider"] === "string" ? body["provider"] : undefined,
        scopes,
      };
    }

    if (resp.status === 401) {
      if (this._onRefresh && attempt < this._maxRetries) {
        // Invoke user-supplied refresh callback and retry with the fresh token.
        this._accessToken = await this._onRefresh();
        return this._exchangeWithRetry(connectionId, attempt + 1);
      }
      throw new VaultError("agent not authorized (401)", 401);
    }

    if (resp.status === 403) {
      throw new VaultError("missing scope for vault access (403)", 403);
    }

    if (resp.status === 404) {
      throw new VaultError(`connection not found: ${connectionId}`, 404);
    }

    throw new VaultError(
      `vault request failed: HTTP ${resp.status}: ${resp.text.slice(0, 200)}`,
      resp.status
    );
  }

  // ------------------------------------------------------------------
  // Admin — disconnect + list connections
  // ------------------------------------------------------------------

  /**
   * Disconnect a vault connection.
   * DELETE /api/v1/admin/vault/connections/{id}
   *
   * @param connectionId The `conn_*` identifier.
   * @param cascade      When true (default) cascade-revokes agent tokens bound to this connection.
   */
  async disconnect(connectionId: string, cascade = true): Promise<VaultDisconnectResult> {
    const qs = cascade ? "?cascade=true" : "?cascade=false";
    const url = `${this._authUrl}/api/v1/admin/vault/connections/${encodeURIComponent(connectionId)}${qs}`;
    const resp = await httpRequest(url, { method: "DELETE", headers: this._adminAuth() });
    if (resp.status === 404) throw new VaultError(`connection not found: ${connectionId}`, 404);
    if (resp.status === 401) throw new VaultError("not authorized (401)", 401);
    if (resp.status === 403) throw new VaultError("forbidden — admin key required (403)", 403);
    if (resp.status !== 200 && resp.status !== 204) {
      throw new VaultError(`disconnect failed: HTTP ${resp.status}: ${resp.text.slice(0, 200)}`, resp.status);
    }
    if (resp.status === 204 || !resp.text) {
      return { connection_id: connectionId, revoked_agent_ids: [], revoked_token_count: 0 };
    }
    const body = resp.json<VaultDisconnectResult>();
    return {
      connection_id: body.connection_id ?? connectionId,
      revoked_agent_ids: body.revoked_agent_ids ?? [],
      revoked_token_count: body.revoked_token_count ?? 0,
      cascade_audit_event_id: body.cascade_audit_event_id,
    };
  }

  /**
   * List vault connections (admin scope).
   * GET /api/v1/vault/connections
   */
  async listConnections(params?: ListConnectionsOptions): Promise<VaultConnectionListResult> {
    const qs = new URLSearchParams();
    if (params?.limit != null) qs.set("limit", String(params.limit));
    if (params?.offset != null) qs.set("offset", String(params.offset));
    if (params?.user_id) qs.set("user_id", params.user_id);
    const query = qs.toString();
    const url = `${this._authUrl}/api/v1/admin/vault/connections${query ? `?${query}` : ""}`;
    const resp = await httpRequest(url, { headers: this._adminAuth() });
    if (resp.status !== 200) {
      throw new VaultError(`listConnections failed: HTTP ${resp.status}: ${resp.text.slice(0, 200)}`, resp.status);
    }
    const raw = resp.json<VaultConnectionListResult | VaultConnection[]>();
    if (Array.isArray(raw)) return { data: raw, total: raw.length };
    return raw as VaultConnectionListResult;
  }

  // ------------------------------------------------------------------
  // fetchToken — preferred name (mirrors Python vault.fetch_token)
  // ------------------------------------------------------------------

  /**
   * Retrieve a fresh access token for the given stored connection.
   *
   * Preferred over {@link exchange} — matches Python SDK's `fetch_token` name.
   * Supports an optional DPoP proof header for DPoP-bound connections.
   *
   * @param referenceToken  Connection ID / reference token identifying the stored connection.
   * @param dpop            Optional DPoP proof header value.
   */
  async fetchToken(referenceToken: string, dpop?: string): Promise<VaultToken> {
    if (!referenceToken) throw new VaultError("referenceToken (connection ID) is required");
    // For DPoP-less requests delegate to the retry-capable private helper.
    if (!dpop) return this._exchangeWithRetry(referenceToken, 0);
    // DPoP path — single attempt (caller manages token refresh for DPoP flows).
    const url = `${this._authUrl}${this._connectionsPath}/${referenceToken}/token`;
    const resp = await httpRequest(url, {
      method: "GET",
      headers: { Authorization: `Bearer ${this._accessToken}`, DPoP: dpop },
    });
    if (resp.status === 200) {
      let body: Record<string, unknown>;
      try { body = resp.json<Record<string, unknown>>(); }
      catch { throw new VaultError(`vault response is not valid JSON: ${resp.text.slice(0, 200)}`); }
      let scopes: string[];
      const rawScopes = body["scopes"] ?? body["scope"];
      if (typeof rawScopes === "string") scopes = rawScopes.split(/\s+/).filter(Boolean);
      else if (Array.isArray(rawScopes)) scopes = rawScopes.map(String);
      else scopes = [];
      return {
        accessToken: String(body["access_token"]),
        expiresAt: typeof body["expires_at"] === "number" ? body["expires_at"] : undefined,
        provider: typeof body["provider"] === "string" ? body["provider"] : undefined,
        scopes,
      };
    }
    if (resp.status === 401) throw new VaultError("agent not authorized (401)", 401);
    if (resp.status === 403) throw new VaultError("missing scope for vault access (403)", 403);
    if (resp.status === 404) throw new VaultError(`connection not found: ${referenceToken}`, 404);
    throw new VaultError(`vault request failed: HTTP ${resp.status}: ${resp.text.slice(0, 200)}`, resp.status);
  }
}
